File: frontend/src/types/scrape.ts
export type DuplicateWarningInfo = {
  id: number;
  title: string;
  url: string;
  image_local_path: string | null;
  title_score: number;
  ingredient_score: number;
};

export type ScrapeResponse = {
  message: string;
  url: string;
  job_id?: number;
  status: "pending" | "running" | "completed" | "failed" | "duplicate_warning";
  similar_recipe?: DuplicateWarningInfo | null;
};

export type JobStatusResponse = {
  job_id: number;
  status: "pending" | "running" | "completed" | "failed" | "duplicate_warning";
  error_message?: string | null;
};

export type RecipeSimilarPair = {
  recipe_a_id: number;
  recipe_a_title: string;
  recipe_a_url: string;
  recipe_a_image: string | null;
  recipe_b_id: number;
  recipe_b_title: string;
  recipe_b_url: string;
  recipe_b_image: string | null;
  title_score: number;
  ingredient_score: number;
};

export type RecipeSimilarPairsResponse = {
  pairs: RecipeSimilarPair[];
};

File: backend/app/schemas/scrape.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, HttpUrl


JobStatus = Literal["pending", "running", "completed", "failed", "duplicate_warning"]


class HFSearchResultResponse(BaseModel):
    id: str
    name: str
    image_url: str | None
    tags: list[str]
    total_time_minutes: int | None
    hf_url: str
    already_imported: bool


class ScrapeRequest(BaseModel):
    url: HttpUrl
    force: bool = False


class DuplicateWarningInfo(BaseModel):
    id: int
    title: str
    url: str
    image_local_path: str | None = None
    title_score: float
    ingredient_score: float


class ScrapeResponse(BaseModel):
    message: str
    url: str
    job_id: int | None = None
    status: JobStatus
    similar_recipe: DuplicateWarningInfo | None = None


class RecipeSimilarPair(BaseModel):
    recipe_a_id: int
    recipe_a_title: str
    recipe_a_url: str
    recipe_a_image: str | None = None
    recipe_b_id: int
    recipe_b_title: str
    recipe_b_url: str
    recipe_b_image: str | None = None
    title_score: float
    ingredient_score: float


class RecipeSimilarPairsResponse(BaseModel):
    pairs: list[RecipeSimilarPair]


class JobStatusResponse(BaseModel):
    job_id: int
    status: JobStatus
    error_message: str | None = None
